packages/astro/src/i18n/index.ts

```diff
@@ -244,3 +276,109 @@ class Unreachable extends Error {
 		);
 	}
 }
+
+export type MiddlewarePayload = {
+	base: string;
+	locales: Locales;
+	trailingSlash: AstroConfig['trailingSlash'];
+	format: AstroConfig['build']['format'];
+	routing?: RoutingStrategies;
+	defaultLocale: string;
+	domains: Record<string, string> | undefined;
+	fallback: Record<string, string> | undefined;
+};
+
+export function redirectToDefaultLocale({
+	trailingSlash,
+	format,
+	base,
+	defaultLocale,
+}: MiddlewarePayload) {
+	return function (context: APIContext, statusCode?: ValidRedirectStatus) {
+		if (shouldAppendForwardSlash(trailingSlash, format)) {
+			return context.redirect(`${appendForwardSlash(joinPaths(base, defaultLocale))}`, statusCode);
+		} else {
+			return context.redirect(`${joinPaths(base, defaultLocale)}`, statusCode);
+		}
+	};
+}
+
+export function notFound({ base, locales }: MiddlewarePayload) {
+	return function (context: APIContext, response?: Response): Response | undefined {
+		if (response?.headers.get(REROUTE_DIRECTIVE_HEADER) === 'no') return response;
+
+		const url = context.url;
+		// We return a 404 if:
+		// - the current path isn't a root. e.g. / or /<base>
+		// - the URL doesn't contain a locale
+		const isRoot = url.pathname === base + '/' || url.pathname === base;
+		if (!(isRoot || pathHasLocale(url.pathname, locales))) {
+			if (response) {
+				response.headers.set(REROUTE_DIRECTIVE_HEADER, 'no');
+				return new Response(null, {
+					status: 404,
+					headers: response.headers,
+				});
+			} else {
+				return new Response(null, {
+					status: 404,
+					headers: {
+						[REROUTE_DIRECTIVE_HEADER]: 'no',
+					},
+				});
+			}
+		}
+
+		return undefined;
+	};
+}
+
+export type UseFallback = (context: APIContext, response: Response) => Response | undefined;
+export function useFallback({ fallback, locales, defaultLocale, routing }: MiddlewarePayload) {
```

Comment:
Question about `useFallback`: The `use` prefix is kinda overloaded by React to mean its React hooks. It seems like what this function does is that it redirects to a fallback page, would it be more accurate as `redirectToFallback` so it's also consistent with `redirectToDefaultLocale`?